### ISSUE TITLE
Use the latest version in gradle-plugin as default

### ISSUE THREAD

**Issue by sschuberth** [arturbosch/detekt#151]
This is a follow-up to [my comment over here](https://github.com/arturbosch/detekt/issues/104#issuecomment-310588921). Using version M12.3, for

```
/**
 * Operating-System-specific utility functions.
 */
object OS {
    private val OS_NAME = System.getProperty("os.name").toLowerCase()

    val isWindows = OS_NAME.contains("windows")
}
```

I get a

    NoDocOverPublicClass - [OS] at src\main\kotlin\Utils.kt:15:1

false positive (on both Windows and Linux). Interestingly, the reported line 15 is the line containing `/**`, *not* the line containing `object OS {`.

**Comment by arturbosch** [comment]
I think your using a old version of detekt. The rule was renamed to `UndocumentedPublicClass`, can this be?

**Comment by sschuberth** [comment]
Hmm, I have

```
plugins {
    id 'io.gitlab.arturbosch.detekt' version '1.0.0.M12.3'
}
```

Do I still need to explicitly set the version in the `detekt` closure, too?

**Comment by arturbosch** [comment]
Uff ... :( the gradle-plugin still uses M11... 

-> you have to specify the version to be up to date. 

I need to find a way to set the latest version as default without hard coding it ...

**Comment by arturbosch** [comment]
I can use `latest.release` as the default version, which gets resolved to the last version of detekt. 

Pros:
- always the newest version
- its default so no worries if I forget to hardcode newest version in gradle-plugin
- user does not need to specify `version`

Cons:
- forced to update/use newest version (build may fail due to new rules -> threshold met)
- need to specify `version` to stay at this version

**Comment by arturbosch** [comment]
975981678fcb20df030952c3d31734630884e192 fixed the versioning problem.

gradle-plugin M12.4 was released.

Pls try if it works for you :)

**Comment by sschuberth** [comment]
Works, thanks a lot for the fast fix!

**Comment by sschuberth** [comment]
Actually, I'm now running into problems with this. When using detekt on CI (breaking the build if the smell threshold is exceeded), it's confusing that the build might break without changing anything just because a new detekt "backend" version was released. Builds need to be deterministic.

Sure, I could simply hard-code the detekt "backend" version in the Gradle closure, but that's inconvenient as I then need to maintain two versions then, the one of the backend and the one of the plugin.

Originally when filing this issue what I mean by "latest version" was "the latest version available by the time the Gradle plugin is released", not "the latest version available by the time the build is executed". I.e. the detekt backend default version should be hard-coded in the Gradle plugin, IMO.

So how about these rules:
- The "backend" and Gradle plugin always have corresponding versions. That is, even is no code was changed in the plugin, but only in the backend, a new version of the plugin is released that hard-codes the new backend version to use.
- The hard-coded backend version can be overridden in the plugin by setting the `version` in the closure. People who want the current behavior can set `version = latest.release` there.

**Comment by sschuberth** [comment]
Ah, I realize my proposal might be a problem because by the time the Gradle plugin is built a new detekt backend version was not yet deployed and cannot be easily used by the plugin mechanism. Maybe this could be solved by [Substituting an external module dependency with a project dependency](https://docs.gradle.org/4.0/userguide/dependency_management.html#sec:module_to_project_substitution), but I didn't want to make things that complicated.

**Comment by arturbosch** [comment]
Yes, that was also my concern as you can see above ^^

To use the same version of the Gradle plugin was the plan, but it's easy to forget to hardcode the newest version. And that was the case from M11 to M12.

My intention was that if you set up a CI you always have to configure more, so hard coding one version more is not as bad as forgetting to increment the hardcode version.

Sometimes there will be a bug in the Gradle plugin, should we then also release a new deteKt version?

Now as I write I think I can just use `project.version` to sync the versions ...

**Comment by arturbosch** [comment]
Project dependency would work, but we loose the version configuration feature. My plan in long run was that the Gradle plugin is version independent of deteKt. 

My though is: there won't be as much updates as for deteKt itself

Synchronizing the two versions will lead to the problem you stated. 

I already wrote a section for new release in the contributing guide, maybe I won't forget to hardcode it anymore. I also can leave the new release changes pull request longer open, so the community can approve it

**Comment by sschuberth** [comment]
> My intention was that if you set up a CI you always have to configure more, so hard coding one version more is not as bad as forgetting to increment the hardcode version.

True. So let's leave things as they are for now. After all it's better to be explicit (by specifying the desired version of the backend in the plugin closure) than implicit (huh, what version is going to be used, esp. once the plugin and backend versions might start to mismatch).
